### PR TITLE
Enable validation function tracing

### PR DESCRIPTION
**Pull request by avpotapov00**

### CODE REVIEW

File: src/jvm/test/resources/expected_logs/validation_function_failure.txt [hunk 1]
```diff
@@ -1,36 +1,56 @@
...
-|  Thread 1   |
-| ----------- |
-| operation() |
-| ----------- |
-| operation() |
-| ----------- |
-| operation() |
-| ----------- |
+| --------------------------------- |
+|      Thread 1       |  Thread 2   |
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/validation_function_failure.txt:3]: It should require only one thread to reproduce the error

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/SpinlockEventsCutTests.kt [hunk 2]
```diff
@@ -226,6 +226,7 @@ class SpinlockInIncorrectResultsWithClocksTest {
                     Actor(method = SpinlockInIncorrectResultsWithClocksTest::d.javaMethod!!, arguments = emptyList())
                 )
             ),
+            emptyList(),
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/SpinlockEventsCutTests.kt:229]: please add argument names

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt [hunk 3]
```diff
@@ -64,9 +65,15 @@ internal fun executeActor(
...
-internal inline fun executeValidationFunctions(instance: Any, validationFunctions: List<Method>,
-                                               onError: (functionName: String, exception: Throwable) -> Unit) {
-    for (f in validationFunctions) {
+internal inline fun executeValidationFunctions(
+    strategy: Strategy,
+    instance: Any,
+    validationFunctions: List<Method>,
+    onError: (functionName: String, exception: Throwable) -> Unit
+) {
+    for (i in validationFunctions.indices) {
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt:74]: You can use `forEachIndexed` instead

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/Strategy.kt [hunk 4]
```diff
@@ -36,4 +36,5 @@ abstract class Strategy protected constructor(
      * Is invoked before each actor execution.
      */
     open fun onActorStart(iThread: Int) {}
+    open fun beforeValidationFunctionCall() {}
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/Strategy.kt:39]: Please add documentation, indicating when this function should be called (and why)

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TraceReporter.kt [hunk 5]
```diff
@@ -33,10 +35,10 @@ internal fun StringBuilder.appendTrace(
 }
 
 private fun StringBuilder.appendShortTrace(
-    startTraceGraphNode: TraceNode?,
+    sectionsFirstNode: List<TraceNode>,
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TraceReporter.kt:38]: `sectionsFirstNode` -- unclear name

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TracePoint.kt [hunk 6]
```diff
@@ -240,6 +241,10 @@ internal class CoroutineCancellationTracePoint(
     }
 }
 
+internal class SectionDelimiterTracePoint(val executionPart: ExecutionPart): TracePoint(0, -1, emptyList()) {
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TracePoint.kt:244]: Why should there be such a trace point? What does "section" mean?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt [hunk 7]
```diff
@@ -85,7 +85,7 @@ abstract class ManagedStrategy(
     // stores a constructor for the corresponding code location.
     private val tracePointConstructors: MutableList<TracePointConstructor> = ArrayList()
     // Collector of all events in the execution such as thread switches.
-    private var traceCollector: TraceCollector? = null // null when `collectTrace` is false
+    internal var traceCollector: TraceCollector? = null // null when `collectTrace` is false
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt:88]: protected?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt [hunk 8]
```diff
@@ -99,8 +99,10 @@ internal class ModelCheckingStrategy(
             ExecutionPart.INIT -> 0
             ExecutionPart.PARALLEL -> currentInterleaving.chooseThread(0)
             ExecutionPart.POST -> 0
+            ExecutionPart.VALIDATION -> 0
         }
         loopDetector.beforePart(nextThread)
+        traceCollector?.passCodeLocation(SectionDelimiterTracePoint(part))
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt:105]: Why do you need this?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt [hunk 9]
```diff
@@ -262,9 +261,26 @@ internal open class ParallelThreadsRunner(
             }
             val afterPostStateRepresentation = constructStateRepresentation()
             // Execute validation functions
-            executeValidationFunctions(testInstance, validationFunctions) { functionName, exception ->
-                return ValidationFailureInvocationResult(scenario, functionName, exception)
+            if (validationFunctions.isNotEmpty()) {
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt:264]: Please use the same scheme as for init/post parts for the validation functions -- use the bytecode generation

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/execution/ExecutionScenario.kt [hunk 10]
```diff
@@ -24,6 +25,7 @@ class ExecutionScenario(
      *
      * The initial part is executed in the first thread of the scenario before the parallel part.
      *
+     *
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/execution/ExecutionScenario.kt:28]: please revert this change

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt [hunk 11]
```diff
@@ -154,6 +155,16 @@ internal class TableLayout(
         appendColumns(data, columnWidths, transform)
     }
 
+    /**
+     * Appends first column.
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt:159]: the first

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt [hunk 12]
```diff
@@ -154,6 +155,16 @@ internal class TableLayout(
...
+     *
+     * @see columnsToString
+     */
+    fun <T> StringBuilder.appendFirstColumn(data: List<T>, transform: ((T) -> String)? = null) = apply {
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt:163]: appendToFirstColumn?

**Review by ndkoval — CHANGES_REQUESTED** [review]

File: src/jvm/test/resources/expected_logs/validation_function_failure.txt (hunk 1 above)
Comment (avpotapov00) [reply src/jvm/test/resources/expected_logs/validation_function_failure.txt:3]: I added the second thread to check that the table is well-formatted in the case when it has more than 2 columns. I added this explanation to the test.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TracePoint.kt (hunk 6 above)
Comment (avpotapov00) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TracePoint.kt:244]: This trace point that is added to the trace between execution parts (init, parallel, post, validation). Added comments to the code.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt (hunk 8 above)
Comment (avpotapov00) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt:106]: To insert sections delimiter trace point. It's convenient to use them lately during report creation. See `SectionDelimiterTracePoint` documentation.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt [hunk 13]
```diff
@@ -346,10 +366,12 @@ internal open class ParallelThreadsRunner(
         clocks = Array(nActors) { emptyClockArray(nThreads) }
     }
 
-    private fun TestThreadExecution.reset() {
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt:349]: Please revert the change

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt [hunk 14]
```diff
@@ -199,22 +210,55 @@ internal class TableLayout(
 internal fun ExecutionLayout(
     initPart: List<String>,
     parallelPart: List<List<String>>,
-    postPart: List<String>
+    postPart: List<String>,
+    validation: List<String>?
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt:214]: Please make the absence of the validation consistent with other execution parts: use emptyList instead of null

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt [hunk 15]
```diff
@@ -199,22 +210,55 @@ internal class TableLayout(
...
     val size = parallelPart.size
     val threadHeaders = (0 until size).map { "Thread ${it + 1}" }
-    val columnsWidth = parallelPart.mapIndexed { i, actors ->
-        val col = actors + if (i == 0) (initPart + postPart) else listOf()
-        col.maxOfOrNull { it.length } ?: 0
+    val columnsWidth = parallelPart.mapIndexed { columnWidth, actors ->
+        var maxColWidth = 0
+        actors.forEach { maxColWidth = max(maxColWidth, it.length) }
+        if (columnWidth == 0) {
+            initPart.forEach { maxColWidth = max(maxColWidth, it.length)  }
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt:222]: initPart + postPart + validationPart

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt [hunk 16]
```diff
@@ -199,22 +210,55 @@ internal class TableLayout(
...
+    postPart: List<String>,
+    validation: List<String>?
 ): TableLayout {
     val size = parallelPart.size
     val threadHeaders = (0 until size).map { "Thread ${it + 1}" }
-    val columnsWidth = parallelPart.mapIndexed { i, actors ->
-        val col = actors + if (i == 0) (initPart + postPart) else listOf()
-        col.maxOfOrNull { it.length } ?: 0
+    val columnsWidth = parallelPart.mapIndexed { columnWidth, actors ->
+        var maxColWidth = 0
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt:219]: Just combine everything and call max()

**Review by ndkoval — CHANGES_REQUESTED** [review]

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt [hunk 17]
```diff
@@ -95,10 +96,12 @@ internal class ModelCheckingStrategy(
...
 
     override fun beforePart(part: ExecutionPart) {
+        super.beforePart(part)
         val nextThread = when (part) {
             ExecutionPart.INIT -> 0
             ExecutionPart.PARALLEL -> currentInterleaving.chooseThread(0)
             ExecutionPart.POST -> 0
+            ExecutionPart.VALIDATION -> 0
         }
         loopDetector.beforePart(nextThread)
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt:106]: Move to super.beforePart(..)

**Review by ndkoval — CHANGES_REQUESTED** [review]

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt (hunk 17 above)
Comment (avpotapov00) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt:106]: I can't do it because `currentInterleaving` is a field of `ModelCheckingStrategy`

**Review by ndkoval — APPROVED** [review]
